cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -357,12 +357,14 @@ Result<ScanTaskIterator> ParquetFileFormat::ScanFile(std::shared_ptr<ScanOptions
                         GetReader(fragment->source(), options.get(), context.get()));
 
   if (!parquet_fragment->HasCompleteMetadata()) {
-    // row groups were not already filtered; do this now
-    RETURN_NOT_OK(parquet_fragment->EnsureCompleteMetadata(reader.get()));
-    ARROW_ASSIGN_OR_RAISE(row_groups,
-                          parquet_fragment->FilterRowGroups(*options->filter));
-    if (row_groups.empty()) {
-      return MakeEmptyIterator<std::shared_ptr<ScanTask>>();
+    // row groups were not already filtered; do this now (if there is a filter)
+    if (!options->filter->Equals(true)) {
```

Comment:
@bkietz ideally we would also skip this if the `flilter` only involves the `partition_expression` and not any actual columns of the file. What is the best way to check this? (simplify the filter first with `partition_expression` ?)

Reply:
Simplifying the filter by the partition expression *would* work, but it's not safe without loading the physical schema since we need to validate the filter against that. Otherwise a filter without implicit casts (for example) would result in an assertion failure. Let's leave that optimization for a follow up; the right way to do this is probably to rewrite `Expression::Assume` to return a Result (allowing it to fail non-catastrophically even without a schema against which to validate), but that's a much larger project